badge-maker/lib/xml.js

```diff
@@ -73,27 +73,24 @@ class XmlElement {
   }
 }
 
-class NullElement {
-  // TODO: we can remove this later
-  render() {
-    return ''
-  }
-}
-
+/**
+ * Convenience class. Sometimes it is useful to return an object that behaves
+ * like an XmlElement but renders multiple XML tags (not wrapped in a <g>).
+ */
 class ElementList {
   constructor({ content = [] }) {
     this.content = content
   }
 
   render() {
-    return this.content.reduce((acc, el) => acc + el.render(), '')
+    return this.content.reduce(
+      (acc, el) =>
+        typeof el.render === 'function'
+          ? acc + el.render()
+          : acc + escapeXml(el),
+      ''
+    )
```

Comment:
I can't convince myself that I understand this part of the diff :laughing: Would you mind explaining it to me?

Reply:
Yep so basically `this.content` is a list of things each of which might be:
- an `XmlElement` object
- another nested `ElementList` object
- a string (tbh, this is mostly an empty string when we don't want to render a particular element e.g: https://github.com/badges/shields/blob/b6c5a10548a2ac9471cf0785f18f7b64c5870bdf/badge-maker/lib/badge-renderers.js#L227-L231 )

so if I say 

```js
const list = new ElementList({content: [
  new XmlElement({name: 'foo'}),
  'bar',
  new XmlElement({name: 'baz'}),
]})
```

then I want `render()` to output

```xml
<foo />bar<baz />
```

---

```js
this.content.reduce((acc, el) => acc + el, '')
```

does the same thing as

```js
this.content.join('')
```

so I'm doing that, but then I'm saying in the reduce function:
If this object has a `render()` function (it quacks like an `XmlElement` or an `ElementList`), call `render()` and concat the result. Otherwise (assume it is just a string), escape the string and concat that.

Reply:
> tbh, this is mostly an empty string when we don't want to render a particular element e.g:

Ah okay, so even if it's not an element it's not strictly going to be an empty string which requires a more thorough check. makes sense, thanks